### PR TITLE
docs: example on Vision API (image to ad copy)

### PR DESCRIPTION
**Pull request by lakshyaag**
This pull request adds the .DS_Store file to the .gitignore file to prevent it from being tracked in the repository. It also includes an example on how to use the GPT-4 Vision API and the Chat API to automatically generate advertising copy from product images.

## Summary by CodeRabbit

- **New Features**
  - Introduced a guide on using GPT-4 Vision API and Chat API to generate ad copy from images.
  - Implemented a new example script for identifying products in images and creating advertising content.

- **Documentation**
  - Added `image_to_ad_copy.md` to showcase usage and capabilities of the new feature.

- **Chores**
  - Updated `.gitignore` to exclude macOS system files.

### CODE REVIEW

**Comment by jxnl** [comment]
Can you include the example output the same way I'd did in the vision post?

File: docs/examples/image_to_ad_copy.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# Use Vision API to detect products and generate advertising copy
```
Comment (jxnl) [inline docs/examples/image_to_ad_copy.md:1]: make sure you update

docs/index.md 
mkdocs.yaml

File: docs/examples/image_to_ad_copy.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    # Can be customized and automatically generated
+    def generate_prompt(self):
+        prompt = f"Product: {self.name}\n"
+        if self.description:
+            prompt += f"Description: {self.description}\n"
+        if self.key_features:
+            prompt += f"Key Features: {', '.join(self.key_features)}\n"
+        return prompt
+
+    def __repr__(self):
```
Comment (jxnl) [inline docs/examples/image_to_ad_copy.md:47]: no need to override repr.

File: docs/examples/image_to_ad_copy.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    headline: str = Field(
+        description="A short, catchy, and memorable headline for the given product. The headline should invoke curiosity and interest in the product.",
+    )
+    ad_copy: str = Field(
+        description="A long-form advertisement copy for the given product. This will be used in campaigns to promote the product with a persuasive message and a call-to-action with the objective of driving sales.",
+    )
+    name: str = Field(
+        description="The name of the product being advertised.",    )
+
+    def __str__(self):
```
Comment (jxnl) [inline docs/examples/image_to_ad_copy.md:96]: no need for __str__

usually model_dumps() is enough

File: docs/examples/image_to_ad_copy.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+```
+
+## Calling the API
+
+### Product Detection
+
+The `read_images` function uses OpenAI's vision model to process a list of image URLs and identify products in each of them. We utilize the `instructor` library to patch the OpenAI client for this purpose.
+
+```python
+def read_images(image_urls: List[str]):
```
Comment (jxnl) [inline docs/examples/image_to_ad_copy.md:110]:
```suggestion
def read_images(image_urls: List[str]) -> IdentifiedProduct:
```

File: examples/vision/image_to_ad_copy.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+    def __str__(self):
+        return f"{self.name}: \n" + "-" * 100 + f"{self.headline}\n{self.ad_copy}"
+
+    def __repr__(self):
+        return str(self)
+
+
+# Define clients
+client_image = instructor.patch(
```
Comment (jxnl) [inline examples/vision/image_to_ad_copy.py:110]: i see, we shuold work on being able to overload the default mode in the create call. in the future

File: examples/vision/image_to_ad_copy.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+    rprint(f"[green]{len(products)} products identified:[/green]")
+    for product, ad_copy in zip(products, ad_copies):
+        rprint(f"[green]{product}[/green]")
+        rprint(f"[blue]Ad Copy: {ad_copy.ad_copy}[/blue]")
+
+    logger.info("Writing results to file...")
+
+    with open("results.json", "w") as f:
+        json.dump(
```
Comment (jxnl) [inline examples/vision/image_to_ad_copy.py:222]: does a single products.model_dump() not work?

File: examples/vision/image_to_ad_copy.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            {
+                "products": [prod.model_dump() for prod in products],
+                "ad_copies": [ad.model_dump() for ad in ad_copies],
+            },
+            f,
+            indent=4,
+        )
```
Comment (jxnl) [inline examples/vision/image_to_ad_copy.py:229]: i try to leave a """ string that shows the outputs

**Comment by jxnl** [comment]
the images don't look like i see a desert and water, not the products

**Comment by lakshyaag** [comment]
> the images don't look like i see a desert and water, not the products

Oh yes, I didn't upload the actual image links because I thought there would be some sort of licensing restrictions.

**Pull request merged**